src/v/cloud_storage/tests/cloud_storage_e2e_test.cc

```diff
@@ -672,3 +673,285 @@ FIXTURE_TEST(
+    // answer it, but won't have a value anyways.
+    make_and_verify_timequery(
+      model::timestamp{
+        base_timestamp() + (batch_time_delta_ms * total_records)},
+      model::offset{total_records},
+      false);
+}
+
+FIXTURE_TEST(test_mixed_timequery, e2e_fixture) {
+    const model::topic topic_name("tapioca");
+    model::ntp ntp(model::kafka_namespace, topic_name, model::partition_id{0});
+
+    // Enable full shadow indexing for now.
+    cluster::topic_properties props;
+    props.shadow_indexing = model::shadow_indexing_mode::full;
+
+    add_topic({model::kafka_namespace, topic_name}, 1, props).get();
+
+    wait_for_leader(ntp).get();
+
+    auto partition = app.partition_manager.local().get(ntp);
+    auto log = partition->log();
+    auto& archiver = partition->archiver().value().get();
+    BOOST_REQUIRE(archiver.sync_for_tests().get());
+
+    // Generate batches [0, 10, 20, ..., 100]
+    const auto num_segs = 11;
+    const auto batches_per_segment = 1;
+    const auto batch_time_delta_ms = 10;
+    tests::remote_segment_generator gen(make_kafka_client().get(), *partition);
+    auto total_records = gen.num_segments(num_segs)
+                           .batches_per_segment(batches_per_segment)
+                           .base_timestamp(model::timestamp{0})
+                           .batch_time_delta_ms(batch_time_delta_ms)
+                           .produce()
+                           .get();
+    BOOST_REQUIRE_EQUAL(total_records, 11);
+
+    const auto base_timestamp = log->start_timestamp();
+    BOOST_REQUIRE_EQUAL(base_timestamp, model::timestamp{0});
+
+    const auto num_segments_to_keep = 2;
+    const auto upper_timestamp = base_timestamp()
+                                 + (num_segs - num_segments_to_keep)
+                                     * batch_time_delta_ms;
+    const auto max_timestamp = base_timestamp()
+                               + (num_segs - 1) * batch_time_delta_ms;
+
+    // Sum the sizes of trailing segments
+    const auto& segments = log->segments();
+    const size_t max_bytes = std::accumulate(
+      std::next(segments.begin(), segments.size() - num_segments_to_keep),
+      segments.end(),
+      size_t{0},
+      [](size_t size, const auto& seg) { return size + seg->file_size(); });
+
+    // Force garbage collection of all local records [0, upper_timestamp). Full
+    // records [0, max_timestamp] still exist in the cloud.
+    storage::gc_config gc_conf(model::timestamp{upper_timestamp}, max_bytes);
+    log->gc(gc_conf).get();
+
+    RPTEST_REQUIRE_EVENTUALLY(10s, [log = partition->log()] {
+        return log->segments().size() == num_segments_to_keep;
+    });
+
+    // Disable remote fetch, forcing local data usage only.
+    auto disable_fetch_override = storage::ntp_config::default_overrides{
+      .shadow_indexing_mode = props.shadow_indexing
+      = model::shadow_indexing_mode::archival};
```

Comment:
nit: I'm confused by this. Why are we setting props.shadow_indexing?

Reply:
The value of `props.shadow_indexing` dictates the result of `partition::is_remote_fetch_enabled()`, since the partition goes into the log config at the raft layer. In `storage/ntp_config.h`, the important check is in `ntp_config::is_remote_fetch_enabled()`, the check being `model::is_fetch_enabled(_overrides->shadow_indexing_mode.value())` in `model/fundamental.h`

Reply:
Sure, but `props` is never used after setting it on this line, so why bother changing its value? Isn't `disable_fetch_override` getting the job done?

Reply:
Oops, I apologize. My eyes definitely overlooked the multiple assignments on this line a number of times. Fixed! :)